src/psi/j/executors/saga.py

```diff
@@ -0,0 +1,222 @@
+import atexit
+import logging
+from threading import RLock
+from typing import Optional, Dict, List
+
+import radical.saga as rs
+import radical.saga.job.constants as ct
+import time
+
+from psi.j import JobExecutor, JobExecutorConfig, JobState, JobStatus, Job, \
+    SubmitException, InvalidJobException
+
+logger = logging.getLogger(__name__)
+
+
+# map SAGA states to jpsi states.
+_STATE_MAP = {
+    rs.job.NEW: JobState.NEW,
+    rs.job.PENDING: JobState.QUEUED,
+    rs.job.RUNNING: JobState.ACTIVE,
+    rs.job.DONE: JobState.COMPLETED,
+    rs.job.FAILED: JobState.FAILED,
+    rs.job.CANCELED: JobState.CANCELED,
+}
+
+
+class _JobMapping:
+    def __init__(self, jpsi_job: Job, saga_job: Optional[rs.job.Job] = None):
+        self.jpsi_job = jpsi_job
+        self.saga_job = saga_job
+
+
+class SagaExecutor(JobExecutor):
+    """
+    An executor that use SAGA to run jobs.
+
+    For details about SAGA, see https://en.wikipedia.org/wiki/Simple_API_for_Grid_Applications
+    and related links.
+    """
+
+    assert('job' in dir(rs)), 'incomplete installation'
+
+    def __init__(self, url: Optional[str] = None,
+                 config: Optional[JobExecutorConfig] = None) -> None:
+        """
+        Initializes a `SagaExecutor`.
+
+        Parameters
+        ----------
+        url
+            An URL representing a SAGA resource manager to connect to.
+        config
+            Not used.
+        """
+        super().__init__(url=url, config=config)
+
+        self._jobs = dict()  # type: Dict[str, _JobMapping]
+        self._lock = RLock()
+        self._js = rs.job.Service(self.url)
+        self._cb = None
+        atexit.register(self._close)
+
+    def _close(self) -> None:
+        if self._js:
+            self._js.close()
```

Comment:
add `self._js = None` to avoid repeated close calls?  Otherwise the `if` will always be `True`

Reply:
True. Either the check is redundant because other things ensure that this is only called once, or it is not, but then what you say.